File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
     void getPermutations(vector<int> &nums, vector<vector<int>> &ans, vector<int> perm, int ind, vector<int> &mapping){
        if(ind == nums.size()){
            ans.push_back(perm);
            return;
        }
         
         for (int i = 0; i < nums.size(); ++i) {
            if (mapping[i] == 0) {
                perm.push_back(nums[i]);
                mapping[i] = 1;
                getPermutations(nums, ans, perm, ind + 1, mapping);
                perm.pop_back();
                mapping[i] = 0;
            }
        }
         
         
         
         return;
     }
    
    void getPermutations(vector<int> &nums, vector<vector<int>> &ans, vector<int> perm, int ind){
        if(ind == nums.size()){
            ans.push_back(perm);
            return;
        }
        
        for(int i=ind; i<nums.size(); i++){
            swap(nums[i], nums[ind]);
            perm.push_back(nums[ind]);
            getPermutations(nums, ans, perm, ind+1);
            swap(nums[i], nums[ind]);
            perm.pop_back();
        }
        
        return ;
    }
    
    vector<vector<int>> permute(vector<int>& nums) {
         vector<vector<int>> ans;
        vector<int> perm;
        vector<int> mapping(nums.size(), 0);
        
        getPermutations(nums, ans, perm, 0, mapping);
        
        return ans;
    }
};

File: 0046-permutations/0046-permutations.java
class Solution {
    public void solve(List<List<Integer>> ans, List<Integer> current, int[] nums, boolean[] freq){
        if(current.size() == nums.length){
            ans.add(new ArrayList(current));
            return;
        }
        for(int i=0; i<nums.length; i++){
            if(freq[i] == false){
                freq[i] = true;
                current.add(nums[i]);
                solve(ans, current, nums, freq);
                current.remove(current.size()-1);
                freq[i] = false;
            }
        }
        return;
    }
    
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> current = new ArrayList<>();
        boolean freq[] = new boolean[nums.length];
        solve(ans, current, nums, freq);
        return ans;
    }
}